### ISSUE TITLE
Locale in the dictionary

### ISSUE THREAD

**Issue by ampli** [opencog/link-grammar#337]
I rewrote the locale setup code for  Windows to solve most of the problems, and I finally realized that in addition to that it will be a good idea to implement what said in this comment from `link-parser.c`:
```
	/* Get the locale from the environment...
	 * Perhaps we should someday get it from the dictionary ??
	 */
```
For dictionaries which doesn't specify a locale, the mechanism will be like now (setting the locale from the environment), so the behavior will be backward compatible.

Regretfully we cannot currently use something like:
```
<dictionary-locale>: ru_RU.UTF-8;
````
Or better:
```
<dictionary-locale>: "u_RU.UTF-8";
```
(To use the last format, struct Dict_node_s needs a field addition, which seems an overkill
for the few variables that can be needed.)

So I propose to use:
```
<dictionary-locale>.ru_RU\.UTF-8: ();
```
I found that a very small code change is needed to support that:
1. When checking for idioms, ignore words starting with **<**.
2. When checking for SUBSCRIPT_DOT,  ignore a **\**-escaped dots.

Storing values in the subscript may seem strange, but since a lookup can be done by a word without a subscript, one maybe can see the subscript also as a part of the data about the word.

(Alternatively it can be defined in4.0.affix.) 

I first thought about defining a locale **list** like `ru_RU.UTF-8:Russian", when the first one whose set succeeds will be used, but for supporting Windows, at least, this is not needed, as it is easy to convert this locale format to values acceptable by Windows (using a system query to convert it - the system knows about format like ru-RU even though setting a locale directly to it doesn't work).

BTW, it seems the locale setup is usually not critical on Linux, since en_US.UTF-8 and almost any other UTF-8 locale (but tr_TR.UTF-8) seem to be appropriate for all languages but Turkish (which in principle must use tr_TR.UTF-8 due to the tolower of `"I"`).
However, on Windows the locale (and the code page which is associated with it by default) is critical for correct keyboard input and tolower.

If this is fine by you I will send pull requests that implement it.

**Comment by linas** [comment]
I'm thinking that maybe putting the locale in the dict is overkill -- how hard is it for the user to just specify this at the shell level, for windows?

But -- if we do put it in the dict, I would rather not modify the dict.  It would be better to encode:
```
<dictionary-locale>:  RUvruvUTFv8+;
```
 so that `RUvruvUTFv8+` is just some crazy connector.  Then look up the `<dictionary-locale>` word, get this connector, and then string-replace the v's by dashes and underbars, to get the locale.

**Comment by ampli** [comment]
There is absolutely no problem on Windows to use getenv("LANG") to get the language name (like "Russian", "English" etc.) which `setlocale()` directly understands. Moreover, there is also no problem to get even formats like `ll_CC` (e.g. he_IL) or even `ll_CC.UTF-8` and translate them internally to the `setlocale()` language name (there is a Windows interface that gets `ll-CC` and outputs a language name suitable to `setlocale()`).  I tested that and intend to include it in any case.

But the locale in the dict has the great benefit of working "from the box" in the **library** level, i.e. there is no need to set the locale by the user or by the programmer, and the locale setting will be always be correct. It is also extremely **easy** to implement it.

The spelling hard-coded directoryname-to-locale can also be eliminated then.
This has the benefit that the user can call the language directory by any name, e.g. en-new, and still get the spelling correction features (in the current lib this would disable spelling). A !locale variable can also be added for users who like another spelling, like en_GB (trivial to implement but not essential).

It is even enough to have only the language name and the country name, because the UTF-8 part is implied in Linux (and in Windows there are no UTF-8 locales anyway).

Note that in Linux the language part **must** be lowercase and the country uppercase (Windows of course doesn't care). So it should be something like:
```
<dictionary-locale>: Lru1RU+;
```
(Cannot have v instead of 1 if we would like to make it general - there are language shortcuts which need 3 letters.) Alternatively, we can use:
```
<dictionary-locale>: RUvRU+;
```
and convert the case as needed.

My goal is to remove the locale setup from link-parser/tests.py etc. since it will be done by the library.

**Comment by linas** [comment]
OK, sounds reasonable to me. For the separator, anything that is visually distinct (eye-catching) will do -- e.g. v or w or any number except 1 or 0 since visually 1 looks like lower-case l and 0 looks like upper-case O.  I'm thinking we should reserve punctuation for "future use" (whatever that may be).

**Comment by ampli** [comment]
I first planned to add library support for setting the locale according to <dictionary-locale>.
However,  the library is multi-threading, and different threads may open different libraries and thus need different locales. This can be done, with some efforts, on POSIX systems, but only with very much effort in Windows.

So I added a UI function `linkgrammar_get_dict_locale()`. If exists, `link-parser` will prefer it over the locale from the environment. (Existing programs can work as before and don't need a change.)

Another issue is supporting setting the locale via the environment on Windows, as `setlocale(...,"")` doesn't do it there. There is a function `win32_getlocale()` in `untilities.c` that do it in order to implement  `dictionary_create_default_lang()` (crashes on `strcpy` to an uninitialized variable, but needs other changes too). The "problem" is that it will need to be duplicated in `link-parser` in order to get the Windows environment there (like `g_fgetc()`).

But in any case, I wonder why `dictionary_create_default_lang()` is needed at all in the UI. It is mentioned in `link-parser.c`, but the only way to call it is to use an empty language argument:
```
link-parser ''
```
This doesn't seem to add functionality...

(BTW, I finished most of the coding in different branches and just need to integrate them.)

**Comment by ampli** [comment]
More on the apparently uneeded `dictionary_create_default_lang() `:
It invokes `get_default_locale()`, which is rather buggy:
```
	locale = NULL;
#ifdef _WIN32
	if(!locale)
		locale = win32_getlocale ();
#endif
	if(!locale)
		locale = safe_strdup (getenv ("LANG"));             <<<<<<< 1
#if defined(HAVE_LC_MESSAGES)
	if(!locale)
		locale = safe_strdup (setlocale (LC_MESSAGES, NULL));
#endif
	if(!locale)
		locale = safe_strdup (setlocale (LC_ALL, NULL));    <<<<<<< 2
	if(!locale || strcmp(locale, "C") == 0) {
		free(locale);
		locale = safe_strdup("en");
	}
```
1. It should check for an empty value, as this is a convention for an unset environment variable. Note also that the environment variable AC_ALL is not checked at all.
2. If a locale has not been set yet in the program, `setlocale(...,NULL)` always returns the locale "C"! On the other hand, if the program has established a particular locale by now, if it used a hard-coded locale then it had no point to call `dictionary_create_default_lang() ` on the first place (which calls the`get_default_locale()` function I discuss here), and if it used the locale from the environment (say from LANG) then this line of code would get skipped. So it has no function in any case.

This function needs a rewrite.

**Comment by linas** [comment]
LG clients can only use one dictionary at a time. This seems like a reasonable restriction at this time. This has nothing to do with threading: even a single-threaded client might try to open two dicts, and parse, alternately with one or with the other.

**Comment by linas** [comment]
One can always invoke the parser by just saying `link-parser` without any arguments. This is different than calling `link-parser ''` with a null argument -- its les typing for the user.

**Comment by linas** [comment]
re `get_default_locale()` -- sure -- fix it.

**Comment by ampli** [comment]
>LG clients can only use one dictionary at a time. This seems like a reasonable restriction at this time. This has nothing to do with threading: even a single-threaded client might try to open two dicts, and parse, alternately with one or with the other.

I don't understand what this limitation of "only use one dictionary at a time" means:

I made a test with 2 threads, each parses concurrently using a different dict (I used the en and ru dicts).
This test was on Linux so they can both work fine with the same locale: en_US.UTF-8.
I noted that the parse results, although interleaved, were just fine.

On Windows, the thread using the Russian dict would need to use the Russian LC_CTYPE, else the results would be incorrect.

**Comment by linas** [comment]
OK, so I guess we can have the statement that, for Windows, clients can only use on dictionary at a time.  

The point is, if you want to run both english and russian parsers, just create two different processes. I see no compelling reason why users might want to run different dictionaries in the same process, either in one, or in more threads.  The fact that it might not work in Windows does not seem like something worth fixing.

**Comment by ampli** [comment]
> One can always invoke the parser by just saying link-parser without any arguments. This is different than calling link-parser '' with a null argument -- its less typing for the user.

Note that currently link-parser behaves differently with no argument and with "":

```
$ export LANG=ru_RU.UTF-8
$ link-parser
/tmp/link-grammar/tests
link-grammar: Info: Dictionary found at ./data/en/4.0.dict          <<<<< en dict
link-grammar: Info: Using locale ru_RU.UTF-8.
link-grammar: Info: Dictionary version 5.3.8.
link-grammar: Info: Library version link-grammar-5.3.8. Enter "!help" for help.
linkparser> ^Dlink-grammar: Info: Freeing dictionary en/4.0.dict
link-grammar: Info: Freeing dictionary en/4.0.affix
Bye.
sh-4.3$ link-parser ""
/tmp/link-grammar/tests
link-grammar: Info: Dictionary found at ./data/ru/4.0.dict          <<<<< ru dict
link-grammar: Info: Using locale ru_RU.UTF-8.
link-grammar: Info: Dictionary version 5.1.0.
link-grammar: Info: Library version link-grammar-5.3.8. Enter "!help" for help.
linkparser> ^Dlink-grammar: Info: Freeing dictionary ru/4.0.dict
link-grammar: Info: Freeing dictionary ru/4.0.affix
Bye.
````

With no argument, link-parser uses a hard coded default "en" as argument to `dictionary_create_lang()`.
The only way to make it use `dictionary_create_default_lang()` is to use an empty argument.

**Comment by linas** [comment]
Huh. OK, well, that's a bug. It should have defaulted to russian, in both cases; that was the intent.

**Comment by ampli** [comment]
According to the following comment a change has been introduced (in 2008) to always use "en" as a default language of link-parser:
`const char     *language="en";  /* default to english, and not locale */`

If I change it back, what should be done if a dictionary according to the language code of the domain is not found? Currently `dictionary_create_default_lang()` just returns NULL, and link-parser just reports:
`fprintf(stderr, "%s: Fatal error: Unable to open dictionary.\n", argv[0]);`

without telling which language was tried, a thing that can be very cryptic. It cannot actually tell that, so it seems to me that `dictionary_create_default_lang()` should print a more detailed error message, like:
`"Cannot find dictionary for language %s"` (probably for verbosity>=1).

**Comment by ampli** [comment]
I just sent PR#339.
```
$ LANG=xx_IL.UTF-8 link-parser
/tmp/link-grammar/deflang
/tmp/link-grammar/deflang/link-parser/.libs/lt-link-parser: Warning: locale ANSI_X3.4-1968 was not UTF-8; force-setting to en_US.UTF-8
link-grammar: Error: Could not open dictionary xx/4.0.dict
link-grammar: Fatal error: Unable to open default dictionary.
```
The bogus notification on locale ANSI_X3.4-1968, as well as the functions `get_default_locale()` and `dictionary_create_default_lang()` will be fixed for POSIX and Windows in the next pull request.
Later I also intend to fix the manual in this regard.

**Comment by ampli** [comment]
Sorry for this long post. I think it may be interesting.

I just finished the per-dictionary-locale change, and it passed preliminary tests on Linux and Windows. More tests are needed, and also some polishing, including adding a test to the test suite.

I started by adding <dictionary-locale> definitions to the "real language" dictionaries (not "any" etc.), and a UI to read it. Then I modified `link-parser` to set the locale from the dict, if it is provided.

However, when finishing this change I noted that the regex compilation is not done using the dictionary locale (because the dictionary locale is set out of the library, only after opening the dict), and this seemed to me very wrong (even though on Linux I don't have an example at hand of a 4.0.regex that matches different strings in different locales).

So I implemented a real per-directory locale, that transparently supports opening several dictionaries at once (using one thread or more). The locale handling is done only in the library. The locale of the program (or thread) is not set any more to match the dictionary. In Windows the program even remains in the initial "C" locale.

The idea in this change is to use the `isw*_l(c, locale)` version of the character classification functions. The locale associated with the dictionary is stored in the `locale_t` new field of the Dictionary structure, and these functions are invoked using it.

But I encountered a problem: 
The function `mbrtowc_l()` doesn't exists in the dynamic glibc (2.22). `mbsrtowcs_l()` could be used for emulating it, but for some reason it exists only in the static glibc!
So in order that `mbrtowc()` will work for any UTF-8 character, a locale with a UTF-8 codeset  is ensured for the program.
 
On Windows `mbrtowc()` / `mbrtowc_l()` cannot be used anyway. The comments in utilities.[ch] refer to that problem (of `mbrtowc()` not working on Windows) as a bug. However, it was a bug to use it on Windows on the first place, because it is used to convert a UTF-8 character to a wide char, and on Windows there is no UTF-8 character sets at all. Moreover, the current code uses **MB_CUR_MAX** as the max number of source characters to convert, and in Windows it is usually **equal** **1** (for all the locales of the languages in the current dicts).

So the pull request for this work will (regretfully) include the changes in the said two phases, when the last phase fixes the first one.

As I said above, I continue with tests and polishing, and maybe will send an additional small pull request for things that can be isolated (like not using UTF-8 for dict definition comparison stuff).I hope to be able to send a pull request for the whole work very soon.

BTW, from the checks I did by now the MSVC compiled binary seems to work fine under Cygwin
but it doesn't know about ^D for existing (it somehow "eats" both ^D and ^Z), so only !exit can be used.

I'm sure that in the process of the said changes, and even maybe before, compiling on Cygwin got broken and I would like to check that too.

**Comment by linas** [comment]
On Tue, Jul 26, 2016 at 4:53 AM, Amir Plivatsky <notifications@github.com>
wrote:

> use "en" as a default language of link-parser:
> const char *language="en"; /* default to english, and not locale */
>

The basic idea is that  "things should just work" for the naive,
unsophisticated user, who is maybe trying link-grammar for the first-time
ever.   I don't want first-time users to read complicated instructions, or
do complicated thinking/debugging, just to run the program for the first
time ever.

Similar remarks apply for the users of the programming API.

As long as we can keep "simple things simple", I don't care about the most
of rest.

**Comment by linas** [comment]
On Tue, Aug 9, 2016 at 1:25 PM, Amir Plivatsky <notifications@github.com>
wrote:

> cause it is used to convert a UTF-8 character to a wide char, and on
> Windows there is no UTF-8 character s


Think of it this way: the dictionaries that we distribute are just "a bunch
of bytes".  The code that reads the dictionaries doesn't care what kind of
bytes they are.   So, what we get into RAM is "just a bunch of bytes". For
windows, we have to convert them into something that Windows knows are
characters, thus the mbrtowcs calls.  On Linux, we know that this "bunch of
bytes" just happen to be valid utf8, and so don't need to do as much.

Anything to simplify this process is an improvement.  The ideal situation
is to treat them as "a bunch of bytes" for as long as possible, without
assuming that they are characters, until we absolutely need to work with
characters.

**Comment by ampli** [comment]
In general, I kept everything compatible to the old behavior in the `link-parser` invocation and the API, but this change:

If you use `link-parser` without arguments, it uses a dictionary according to your locale (I think we already agreed these was the intention), unless there is no such dictionary, in that case it defaults to "en".
The user of course can specify any directory as argument. **No debug is needed for all of that.** It may be needed only in the cases that the **current** `link-parser` doesn't work fine due to a locale problem. If this way of locating the dict is not fine, we can of course discuss again how to do it in a fine way.

For API uses, you can still use dictionaries without <dictionary-locale>, as before. You just need to set the appropriate locale before reading it. And if you use one of the current dictionaries on Windows, you don't need complicated setup of input and output code pages and locale in ahead for any language which is not English (if you system language is English) and for English (if your system language is not English).

Currently the naive users (and also not so naive) has no ability to get any dict processing right on Windows (more on the reasons for that - in my next post).

And note that currently tr_TR.UTF-8 users will get a **malfunctioned** English processing, and users which use any other locale will get a malfunctioned "tr" dictionary processing (supposing it could parse something useful).

In addition, if one would like to use locale-dependent regex stuff, it is critical to set the locale correctly per the dict.

**Comment by linas** [comment]
Ugh. Why would setting the locale to `tr_TR.UTF-8` cause English to malfunction?  The whole point of using utf8 is that its almost completely locale-independent... (since we don't use LC_COLLATE LC_TIME LC_NUMERIC etc. and everything else should be the same, no matter what the locale

**Comment by ampli** [comment]
> Think of it this way: the dictionaries that we distribute are just "a bunch
of bytes".  The code that reads the dictionaries doesn't care what kind of
bytes they are.   So, what we get into RAM is "just a bunch of bytes". For
windows, we have to convert them into something that Windows knows are
characters, thus the mbrtowcs calls.  On Linux, we know that this "bunch of
bytes" just happen to be valid utf8, and so don't need to do as much.

Your note here on **Windows** is not exact. **On Windows you cannot use mbrtowcs() for that at all!**

This is because (as I already mentioned above) you actually need to convert UTF-8 codeset input to wide char. However, **Windows** just **doesn't have** any concept of UTF-8 codeset, character set, or locale.... So on **Windows**  MultiByteToWideChar(CP_UTF8, ...) **must be** used instead of `mbrtowcs()`.

Moreover, on **Windows** you just cannot refer to the input text as "just a bunch of bytes".
This is because `iswlower()`/`isupper()` there is locale dependent! I.e. If the locale is not Russian, for example, then `iswupper()` un uppercase Russian letters always returns **FALSE**!!! And the like for the other locales. So on **Windows** (at least) it is absolutely critical to set the correct locale for the dictionary. Note that in the current library even the **Windows** keyboard reading is locale-dependent (i.e. garbage will be read if your locale is not appropriate).

So my patch for per-dict-locale absolutely fixes all locale-setting problems on **Windows** - absolutely no locale setting is needed due to this fix!!! Everything works "from the box" for the naive user!

**Comment by ampli** [comment]
> Ugh. Why would setting the locale to tr_TR.UTF-8 cause English to malfunction? The whole point of using utf8 is that its almost completely locale-independent... (since we don't use LC_COLLATE LC_TIME LC_NUMERIC etc. and everything else should be the same, no matter what the locale

Try this:
```
$ LANG=tr_TR.utf-8 link-parser en
...
linkparser> Is it fine?
No complete linkages found.
Found 1 linkage (1 had no P.P. violations) at null count 1
	Unique linkage, cost vector = (UNUSED=1 DIS= 1.20 LEN=6)

    +-----------Xp----------+
    +---------Wa--------+   |
    |       +-----AN----+   |
    |       |           |   |
LEFT-WALL Is[!] [it] fine.n ? 
```

This is because the downcasing rule for the letter `I` is different than in English... (it is downcased to a dotless i). A similar problem would happen to with non-TR locales and the tr dict.

Regarding "everything else should be the same," note that definitions in 4.0.regex like [:alnum:] and [:punct:] may be locale dependent (in Windows at least I observed parse differences due to that, but this is a totally different issue that we can defer to later).

**Comment by ampli** [comment]
To sum up my per-dict-locale patch:

1. It absolutely solves all locale problems on **Windows**. No locale/code page setup is needed by the user any more. **This is the main issue of this patch** - to make LG work on Windows with no problems.

For this purpose I opened this issue - to make the Windows work trivial and seamless!

On Windows it also solves several other problems (like UTF-8 total input and output transparently that is not locale-dependent), reading batch file, and writing output to files.

2. On other systems it solves the tr_TR.utf-8 vs other locales confusion. Again, no setup is needed by the user, and the user doesn't need to be aware of this problem.

3. It hides the locale setup stuff from the API user. No need at all any more to set up the initial locale to a UTF-8 one (on Linux) or to the particular dict locale (on Windows).

4. It fixes several other problems in link-parser. It includes a big simplification of the input code.
The special Mac fgetc/ungetc stuff is removed (as getc/fgetc is not used any more, while preserving the exact same functionality). The locale handling is totally removed from it (as it is in the  lib).

5. Many source cleanups are done (which have not been included in the pull requests I have already sent).

My question is what to do next.

**Comment by linas** [comment]
Ah, OK, Now I understand ...  except for this:

> My question is what to do next.

I dunno, create a pull request? I skimmed this thread, and didn't spot any other discussions/questions that needed to be resolved.

**Comment by ampli** [comment]
I propose to add to Changelog:
```
* Pull req #350: Add dictionary locale definition.
                       Set the dictionary locale in the library.
                       Improve Windows MSVC/Cygwin/MinGW support.
```
If this is fine I will send a PR for that.

**Comment by linas** [comment]
Is the work on locale-in-dict finished?

**Comment by linas** [comment]
anyway, I just updated the changelog.  I'm thinking a new release tonight or tomorrow.

**Comment by ampli** [comment]
> Is the work on locale-in-dict finished?

Yes.

I still would like to make the spell checkers use it if available instead of the hardwired dict-name
to ll_CC locale conversion. But this can be postponed to 5.3.9.

Note that this work also included correct compilation and working of Cygwin and MinGW and MSVC. This work is not complete yet due to the following:

1. MinGW still cannot print UTF-8 on windows console (e.g. under cmd.exe). Under Cygwin MinGW binary prints UTF-8 fine. The reason is described in PR #352. I still investigate that.
2. MinGW under Cygwin cannot currently produce Python bindings because it tries to use Cygwin's Python.
3. I haven't tried yet MinGW under MSYS.
4. I don't know yet (because I have no time to try to find) the correct PYTHONPATH for the produced Python bindings.
5. As you know, Python bindings is missing on Windows MSVC. I still need to make it.
6. I didn't check the Java bindings on any of MSVC/Cygwin/MinGW (and even not on Linux...).
7. I have a mingw/README detailed file that I have not included, because it still has a problem as described in (1) above. Maybe I will be able to fix it in the next fey hours.

Of course it is best not to delay 5.3.8 due to any of the above.

**Comment by linas** [comment]
ok thanks

**Comment by linas** [comment]
Closing this, because local-in-dict work is done...
